pkg/plugins/builtin/batch/batch.go

```diff
@@ -0,0 +1,362 @@
+	if err != nil {
+		return BatchMetadata{}, err
+	}
+
+	taskIDs, err := populateBatch(ctx, b, dbp, conf, batchCtx)
+	if err != nil {
+		return BatchMetadata{}, err
+	}
+
+	return BatchMetadata{
+		BatchID:        b.PublicID,
+		RemainingTasks: int64(len(conf.Inputs)),
+		TasksStarted:   int64(len(taskIDs)),
+	}, nil
+}
+
+// populateBatch spawns new tasks in the batch and returns their public identifier.
+func populateBatch(
+	ctx context.Context,
+	b *task.Batch,
+	dbp zesty.DBProvider,
+	conf *BatchConfig,
+	batchCtx *BatchContext,
+) ([]string, error) {
+	tasksStarted := batchCtx.metadata.TasksStarted
+	running, err := batchutils.RunningTasks(dbp, b.ID)
+	if err != nil {
+		return []string{}, err
+	}
+
+	// Computing how many tasks to start
+	remaining := int64(len(conf.Inputs)) - tasksStarted
+	toStart := int64(conf.SubBatchSize) - running // How many tasks can be started
+	if remaining < toStart {
+		toStart = remaining // There's less tasks to start remaining than the amount of available running slots
+	}
+
+	args := batch.TaskArgs{
+		TemplateName:     conf.TemplateName,
+		CommonInput:      conf.CommonInput,
+		Inputs:           conf.Inputs[tasksStarted : tasksStarted+toStart],
+		Comment:          conf.Comment,
+		WatcherGroups:    conf.WatcherGroups,
+		WatcherUsernames: conf.WatcherUsernames,
+		Tags:             conf.Tags,
+	}
+
+	taskIDs, err := batch.Populate(ctx, b, dbp, args)
+	if err != nil {
+		return []string{}, err
+	}
+
+	return taskIDs, nil
+}
+
+// runBatch runs batch, spawning new tasks if needed and checking whether they're all done.
+func runBatch(
+	ctx context.Context,
+	conf *BatchConfig,
+	batchCtx *BatchContext,
+	dbp zesty.DBProvider,
+) (BatchMetadata, error) {
+	metadata := batchCtx.metadata
+
+	b, err := task.LoadBatchFromPublicID(dbp, metadata.BatchID)
+	if err != nil {
+		if jujuErrors.IsNotFound(err) {
+			// The batch has been collected (deleted in DB) because no remaining task referenced it. There's
+			// nothing more to do.
+			return metadata, nil
+		}
+		return metadata, err
+	}
+
+	if metadata.TasksStarted < int64(len(conf.Inputs)) {
+		// New tasks still need to be added to the batch
+
+		taskIDs, err := populateBatch(ctx, b, dbp, conf, batchCtx)
+		if err != nil {
+			return metadata, err
+		}
+
+		started := int64(len(taskIDs))
+		metadata.TasksStarted += started
+		metadata.RemainingTasks -= started // Starting X tasks means that X tasks became DONE
+		return metadata, nil
+	}
+	// else, all tasks are started, we need to wait for the last ones to become DONE
+
+	running, err := batchutils.RunningTasks(dbp, b.ID)
+	if err != nil {
+		return metadata, err
+	}
+	metadata.RemainingTasks = running
+	return metadata, nil
+}
+
+// increaseRunMax increases the maximum amount of runs of the resolution matching the given parentTaskID by the run
+// count of the given batchStepName.
+// Since child tasks wake their parent up when they're done, the resolution's RunCount gets incremented everytime. We
```

Comment:
```suggestion
// Since child tasks wake up their paren when they're done, the resolution's RunCount gets incremented everytime. We
```

Reply:
I'm pretty sure "waking [someone / something] up" is correct. :upside_down_face:
Though, both seem fine: https://dictionary.cambridge.org/fr/dictionnaire/anglais/wake-up